src/analysis/processing/qgsalgorithmrandompointsextent.cpp

```diff
@@ -0,0 +1,180 @@
+}
+
+QString QgsRandomPointsExtentAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm creates a new point layer with a given "
+                      "number of random points, all of them within a given extent. "
+                      "A distance factor can be specified, to avoid points being "
+                      "too close to each other. If the minimum distance between points "
+                      " makes it impossible to create new points, either "
+                      " distance can be decreased or the maximum number of attempts may be "
+                      " increased."
+                    );
+}
+
+QgsRandomPointsExtentAlgorithm *QgsRandomPointsExtentAlgorithm::createInstance() const
+{
+  return new QgsRandomPointsExtentAlgorithm();
+}
+
+bool QgsRandomPointsExtentAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  mCrs = parameterAsCrs( parameters, QStringLiteral( "CRS" ), context );
+  mExtent = parameterAsExtent( parameters, QStringLiteral( "EXTENT" ), context, mCrs );
+  mNumPoints = parameterAsInt( parameters, QStringLiteral( "NUM_POINTS" ), context );
+  mDistance = parameterAsDouble( parameters, QStringLiteral( "MIN_DISTANCE" ), context );
+  mMaxAttempts = parameterAsInt( parameters, QStringLiteral( "MAX_ATTEMPTS" ), context );
+
+  return true;
+}
+
+QVariantMap QgsRandomPointsExtentAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  QgsFields fields = QgsFields();
+  fields.append( QgsField( QStringLiteral( "id" ), QVariant::LongLong ) );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields, QgsWkbTypes::Point, mCrs ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  //initialize random engine
+  std::random_device random_device;
+  std::mt19937 mersenne_twister( random_device() );
+
+  std::uniform_real_distribution<double> x_distribution( mExtent.xMinimum(), mExtent.xMaximum() );
+  std::uniform_real_distribution<double> y_distribution( mExtent.yMinimum(), mExtent.yMaximum() );
+
+  if ( mDistance == 0 )
+  {
+    int i = 0;
+    while ( i < mNumPoints )
+    {
+      if ( feedback && feedback->isCanceled() )
+        break;
+
+      double rx = x_distribution( random_device );
+      double ry = y_distribution( random_device );
+
+      QgsFeature f = QgsFeature( i );
+
+      f.setGeometry( QgsGeometry( new QgsPoint( rx, ry ) ) );
+      f.setAttributes( QgsAttributes() << i );
+      sink->addFeature( f, QgsFeatureSink::FastInsert );
+      i++;
+      feedback->setProgress( static_cast<int>( static_cast<double>( i ) / static_cast<double>( mNumPoints ) * 100 ) );
+    }
+  }
+  else
+  {
+    std::unique_ptr<QgsSpatialIndex> index( new QgsSpatialIndex() );
```

Comment:
No need to make this on the heap:

```suggestion
    QgsSpatialIndex index;
```